### PR TITLE
Make #link_to_user polymorphic for CourseUsers

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

File: app/helpers/course/controller_helper.rb [hunk 1]
```diff
@@ -27,4 +27,23 @@ def link_to_course_user(user, options = {})
...
+  # @param [CourseUser|User] user The User to display.
+  # @param [Hash] options The options to pass to +link_to+
+  # @param [Proc] block The block to use for displaying the user.
+  # @return [String] The user-visible string, including embedded HTML which will display the
+  #   string within a link to bring to the User page.
+  def link_to_user(user, options = {}, &block)
+    course = controller.try(:current_course)
+    if user.is_a?(CourseUser)
+      link_to_course_user(user, options, &block)
+    elsif course && course_user = course.course_users.find(user: user)
```
Comment (lowjoel) [inline app/helpers/course/controller_helper.rb]: find will cause unsuccessful lookups to raise an exception?

File: spec/helpers/course/controller_helper_spec.rb [hunk 2]
```diff
@@ -3,6 +3,10 @@
 RSpec.describe Course::ControllerHelper do
   let(:instance) { create(:instance) }
   with_tenant(:instance) do
+    before(:all) do
+      # This is to fix https://github.com/rspec/rspec-rails/issues/1483
+      Course::ControllerHelper.include ApplicationHelper
+    end
     describe '#display_course_user' do
```
Comment (lowjoel) [inline spec/helpers/course/controller_helper_spec.rb]: Add an extra newline before

**Comment by lowjoel** [comment]
Add the github keyword for closing the issue.

File: app/helpers/course/controller_helper.rb [hunk 3]
```diff
@@ -27,4 +27,23 @@ def link_to_course_user(user, options = {})
...
+
+  # Links to the given User or CourseUser. If a User is given, the CourseUser under
+  # current_course of the given user will be displayed.
+  #
+  # @param [CourseUser|User] user The User to display.
+  # @param [Hash] options The options to pass to +link_to+
+  # @param [Proc] block The block to use for displaying the user.
+  # @return [String] The user-visible string, including embedded HTML which will display the
+  #   string within a link to bring to the User page.
+  def link_to_user(user, options = {}, &block)
```
Comment (allenwq) [inline app/helpers/course/controller_helper.rb]: Does this make sense to you ? Basically I think we need to call `super` and link to the User when course_user cannot be found. (i.e. when admin creates something, there is no course_user)

Comment (allenwq) [inline app/helpers/course/controller_helper.rb]: @lowjoel

**Comment by allenwq** [comment]
I think I added the keyword ?

File: app/helpers/course/controller_helper.rb (hunk 1 above)
Comment (allenwq) [inline app/helpers/course/controller_helper.rb]: Oh, yes ! And I haven't tested this line yet, put here for discussing first.

File: app/helpers/course/controller_helper.rb (hunk 3 above)
Comment (lowjoel) [inline app/helpers/course/controller_helper.rb]: I assume that the application helper has the same signature, and if so, yes this should fall back.

**Comment by lowjoel** [comment]
ah, haha I didn't see.

File: app/helpers/course/controller_helper.rb [hunk 4]
```diff
@@ -27,4 +27,23 @@ def link_to_course_user(user, options = {})
...
+  # Links to the given User or CourseUser. If a User is given, the CourseUser under
+  # current_course of the given user will be displayed.
+  #
+  # @param [CourseUser|User] user The CourseUser/User to display.
+  # @param [Hash] options The options to pass to +link_to+
+  # @param [Proc] block The block to use for displaying the user.
+  # @return [String] The user-visible string, including embedded HTML which will display the
+  #   string within a link to bring to the User page.
+  def link_to_user(user, options = {}, &block)
+    course = controller.try(:current_course)
```
Comment (lowjoel) [inline app/helpers/course/controller_helper.rb]: actually no need to try, this helper is only used with Course::Controller, which has current_course.

File: app/helpers/course/controller_helper.rb [hunk 5]
```diff
@@ -27,4 +27,23 @@ def link_to_course_user(user, options = {})
...
+    if user.is_a?(CourseUser)
+      link_to_course_user(user, options, &block)
+    elsif course && course_user = course.course_users.find_by(user: user)
```
Comment (lowjoel) [inline app/helpers/course/controller_helper.rb]: Can you rewrite without using the assignment in the conditional?

File: app/views/course/forum/topics/_topic.html.slim [hunk 6]
```diff
@@ -20,8 +20,4 @@
       = link_to last_post.title, course_forum_topic_path(current_course, @forum, topic)
       br
       = t('.creator')
-      - last_post_creator = current_course.course_users.where(user: last_post.creator)
-      - if last_post_creator
-        = link_to_user(last_post_creator)
-      - else
-        = link_to_user(last_post.creator)
+      = link_to_user(last_post.creator)
```
Comment (lowjoel) [inline app/views/course/forum/topics/_topic.html.slim]: Is this the only place we use this..?

File: app/helpers/course/controller_helper.rb (hunk 4 above)
Comment (allenwq) [inline app/helpers/course/controller_helper.rb]: In helper specs it doesn't have such function, should I always define a current_course in specs ?

Comment (lowjoel) [inline app/helpers/course/controller_helper.rb]: then you should stub it in the spec.

Comment (allenwq) [inline app/helpers/course/controller_helper.rb]: Yes, thanks !

File: app/views/course/forum/topics/_topic.html.slim (hunk 6 above)
Comment (allenwq) [inline app/views/course/forum/topics/_topic.html.slim]: Find one more place ><

**Comment by allenwq** [comment]
@lowjoel  Thanks for renaming !

File: app/views/course/forum/topics/_topic.html.slim (hunk 6 above)
Comment (lowjoel) [inline app/views/course/forum/topics/_topic.html.slim]: Can grep for link_to_user...

Comment (allenwq) [inline app/views/course/forum/topics/_topic.html.slim]: That's what I did just now, only these 2 places

Comment (weiqingtoh) [inline app/views/course/forum/topics/_topic.html.slim]: If anything, this will *probably* come in handy when I do the user badge.

**Comment by lowjoel** [comment]
beautiful

**Comment by allenwq** [comment]
lol, thanks !

**Pull request merged**